cachito/web/models.py

```diff
@@ -890,31 +864,22 @@ def _state_to_json(state):
             latest_state = states[0]
             rv["state_history"] = states
 
-            package_to_deps = {}
-            # This is used to quickly see if a dependency has been added to the root "dependencies"
-            # array in the returned JSON. This uses a set instead of iterating through the
-            # "dependencies" array to avoid an O(n) operation for every dependency.
-            seen_dependencies = set()
+            rv["packages"] = []
             rv["dependencies"] = []
-            for req_dep in self.request_dependencies:
-                dep_json = req_dep.dependency.to_json(
-                    req_dep.replaced_dependency.to_json() if req_dep.replaced_dependency else None,
-                    force_replaces=True,
-                )
-                # Avoid duplicate dependencies in the root "dependencies" array
-                if req_dep.dependency.id not in seen_dependencies:
-                    rv["dependencies"].append(dep_json)
-                    seen_dependencies.add(req_dep.dependency.id)
 
-                package_to_deps.setdefault(req_dep.package.id, []).append(dep_json)
+            bundle_dir = RequestBundleDir(
+                self.id, root=flask.current_app.config["CACHITO_BUNDLES_DIR"]
+            )
+            if bundle_dir.packages_data.exists():
+                packages_data = json.loads(bundle_dir.packages_data.read_bytes())
+                packages = packages_data.get("packages")
+                if packages:
+                    rv["packages"] = packages
+                    deps = []
+                    for package in packages_data["packages"]:
+                        deps.extend(package.get("dependencies", []))
+                    rv["dependencies"] = deps
```

Comment:
The dependencies will need to be de-duplicated by `(name, version, type, dev)`

It would also be good to add unit tests where 2 different packages have the same dependency

Reply:
I think it's also worth mentioning that the dependencies will need to be sorted here as well (once #481 is merged, that is).